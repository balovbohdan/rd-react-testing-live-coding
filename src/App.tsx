import { useState, createContext, useContext, useEffect, useCallback } from 'react';
import { v4 } from 'uuid';
import logo from './logo.svg';
import './App.css';

/*
App
  Header1
    Header2
      Paragraph1
        Paragraph2
*/

const CounterContext = createContext<{
  value: number,
  incerement: () => void
}>({
  value: 0,
  incerement() {}
});

const useCounterContext = () =>
  useContext(CounterContext);

function Paragraph() {
  const counterContext = useCounterContext();

  return (
    <p>
        <img src={logo} className="App-logo" alt="logo" />
        Edit <code>src/App.tsx</code> and save to reload.
        Count: {counterContext.value}
        <button onClick={() => {
          counterContext.incerement();
        }}>Increment</button>
      </p>
  )
}

function Header() {
  return (
    <header className="App-header">
      <Paragraph />
      <a
        className="App-link"
        href="https://reactjs.org"
        target="_blank"
        rel="noopener noreferrer"
      >
        Learn React
      </a>
    </header>
  );
}

function App() {
  const [count, setCount] = useState<number>(0);

  const handleDocumentClick = useCallback(() => {
    console.log('document clicked!');
  }, []);

  useEffect(() => {
    document.addEventListener('click', handleDocumentClick);

    return () => {
      document.removeEventListener('click', handleDocumentClick);
    };
  }, [handleDocumentClick]);

  return (
    <CounterContext.Provider value={{
      value: count,
      incerement() {
        setCount((currentCount) => currentCount + 1);
      }
    }}>
      <div className="App">
        <Header />
        {v4()}
      </div>
    </CounterContext.Provider>
  );
}

export default App;
