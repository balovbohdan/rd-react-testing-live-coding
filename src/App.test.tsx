import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';

const addEventListener = jest.spyOn(document, 'addEventListener');

jest.mock('uuid', () => ({
  v4: () => 'uuid_v4'
}));

beforeEach(() => {
  jest.clearAllMocks();
});

test('it renders', () => {
  const { container } = render(<App />);

  expect(container).toMatchSnapshot();
});

test('it increments (fireEvent)', () => {
  const { debug } = render(<App />);

  const button = screen.getByRole('button', { name: /increment/i });

  userEvent.click(button);
  userEvent.click(button);
  userEvent.click(button);

  // eslint-disable-next-line testing-library/no-debugging-utils
  debug();
});

test('id adds document click listener', async () => {
  render(<App />);

  await waitFor(() => {
    expect(addEventListener).toBeCalledTimes(2);
  });
});

test('renders learn react link', () => {
  render(<App />);
  const linkElement = screen.getByText(/learn react/i);
  expect(linkElement).toBeInTheDocument();
});
