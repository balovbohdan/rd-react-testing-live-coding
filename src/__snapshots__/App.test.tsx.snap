// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <p>
        <img
          alt="logo"
          class="App-logo"
          src="logo.svg"
        />
        Edit 
        <code>
          src/App.tsx
        </code>
         and save to reload. Count: 
        0
        <button>
          Increment
        </button>
      </p>
      <a
        class="App-link"
        href="https://reactjs.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        Learn React
      </a>
    </header>
    uuid_v4
  </div>
</div>
`;
